Fix lvg module read issue with symbolic link on dev

Erouan50 opened the pull request:
Fix #43159

##### SUMMARY
As explained on the issue, the module lvg can't read the physical
volume correctly from a device with a symbolic link if it already
exists. The problem, I guess, is that the comment line 157 of the module
file is wrong. pvcreate uses the "finale" name of the device from udev
rules and not the original one. Example:

Let's assume that I have got a new instance with a secondary EBS volume
named /dev/sdb as a symlink to /dev/nvme1n1. If I want to create a new
physical volume, I should execute that:

```bash
sudo pvcreate /dev/sdb
  Physical volume "/dev/sdb" successfully created.
sudo pvdisplay
  "/dev/sdb" is a new physical volume of "8.00 GiB"
  --- NEW Physical volume ---
  PV Name               /dev/sdb
  VG Name
  PV Size               8.00 GiB
  Allocatable           NO
  PE Size               0
  Total PE              0
  Free PE               0
  Allocated PE          0
  PV UUID               hn17WV-wcKs-0QcT-iqTD-WfbE-22p2-JQm4HQ
```

We can see that the name is still /dev/sdb and not /dev/nvme1n1 in
contrary to the comment. And if the "original" name is used, here is the
result:

```bash
sudo pvcreate /dev/nvme1n1
  Physical volume "/dev/nvme1n1" successfully created.
sudo pvdisplay
  "/dev/sdb" is a new physical volume of "8.00 GiB"
  --- NEW Physical volume ---
  PV Name               /dev/sdb
  VG Name
  PV Size               8.00 GiB
  Allocatable           NO
  PE Size               0
  Total PE              0
  Free PE               0
  Allocated PE          0
  PV UUID               y2UyOK-qey0-rF1A-V0mf-cplg-9nqr-6yQoRd
```

As you can see, the PV name is /dev/sdb instead of /dev/nvme1n1.

So I propose to remove the comment and its associated code to use
directly the name passed by the task. The only problem with that is if
someone uses the /dev/nvme1n1 as device name instead of its symlink, the
issue will be the same. But in both cases, the point is the same if we
change nothing.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
lvg

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (fix-lvg d556e6c79a) last updated 2018/07/23 14:30:28 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/erouan/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/erouan/Documents/ansible/lib/ansible
  executable location = /home/erouan/Documents/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

Erouan50 committed:
Fix lvg module read issue with symbolic link on dev

Fix #43159

As explained on the issue, the module lvg can't read the physical
volume correctly from a device with a symbolic link if it already
exists. The problem, I guess, is that the comment line 157 of the module
file is wrong. pvcreate uses the "finale" name of the device from udev
rules and not the original one. Example:

Let's assume that I have got a new instance with a secondary EBS volume
named /dev/sdb as a symlink to /dev/nvme1n1. If I want to create a new
physical volume, I should execute that:

```bash
sudo pvcreate /dev/sdb
  Physical volume "/dev/sdb" successfully created.
sudo pvdisplay
  "/dev/sdb" is a new physical volume of "8.00 GiB"
  --- NEW Physical volume ---
  PV Name               /dev/sdb
  VG Name
  PV Size               8.00 GiB
  Allocatable           NO
  PE Size               0
  Total PE              0
  Free PE               0
  Allocated PE          0
  PV UUID               hn17WV-wcKs-0QcT-iqTD-WfbE-22p2-JQm4HQ
```

We can see that the name is still /dev/sdb and not /dev/nvme1n1 in
contrary to the comment. And if the "original" name is used, here is the
result:

```bash
sudo pvcreate /dev/nvme1n1
  Physical volume "/dev/nvme1n1" successfully created.
sudo pvdisplay
  "/dev/sdb" is a new physical volume of "8.00 GiB"
  --- NEW Physical volume ---
  PV Name               /dev/sdb
  VG Name
  PV Size               8.00 GiB
  Allocatable           NO
  PE Size               0
  Total PE              0
  Free PE               0
  Allocated PE          0
  PV UUID               y2UyOK-qey0-rF1A-V0mf-cplg-9nqr-6yQoRd
```

As you can see, the PV name is /dev/sdb instead of /dev/nvme1n1.

So I propose to remove the comment and its associated code to use
directly the name passed by the task. The only problem with that is if
someone uses the /dev/nvme1n1 as device name instead of its symlink, the
issue will be the same. But in both cases, the point is the same if we
change nothing.

Erouan50 commented:
Well I just figure out that we might be able to use udevadm info to get the right symlink:
```bash
$ udevadm info --query=symlink --name=/dev/sdb
disk/by-id/lvm-pv-uuid-KOGoxI-aiDA-Ki35-34fg-67if-ozxn-pdgWP0 disk/by-id/nvme-Amazon_Elastic_Block_Store_vol01d498a37cbf542dd disk/by-id/nvme-Amazon_Elastic_Block_Store_vol01d498a37cbf542dd-ns-1 disk/by-id/nvme-nvme.1d0f-766f6c3031643439386133376362663534326464-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001 disk/by-path/pci-0000:00:1f.0-nvme-1 sdb # <----
$ udevadm info --query=symlink --name=/dev/nvme1n1
disk/by-id/lvm-pv-uuid-KOGoxI-aiDA-Ki35-34fg-67if-ozxn-pdgWP0 disk/by-id/nvme-Amazon_Elastic_Block_Store_vol01d498a37cbf542dd disk/by-id/nvme-Amazon_Elastic_Block_Store_vol01d498a37cbf542dd-ns-1 disk/by-id/nvme-nvme.1d0f-766f6c3031643439386133376362663534326464-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001 disk/by-path/pci-0000:00:1f.0-nvme-1 sdb # <----
```
I'm not an expert of lvm and udev. Should I replace what I deleted by this query?

NewJorg commented:
@Erouan50
I have the following setup:
```
root@rescue ~ # ls -l /dev/mapper/decrypted
lrwxrwxrwx 1 root root 7 Aug 26 22:34 /dev/mapper/decrypted -> ../dm-0
root@rescue ~ # pvdisplay /dev/dm-0
  --- Physical volume ---
  PV Name               /dev/mapper/decrypted
  VG Name               pandora
…
root@rescue ~ # pvdisplay /dev/mapper/decrypted 
  --- Physical volume ---
  PV Name               /dev/mapper/decrypted
  VG Name               pandora
…
```

With your fix it is working, when I give the PV Name
```
- lvg vg=pandora pvs=/dev/mapper/decrypted
```

But it fails if I give the real path
```
- lvg vg=pandora pvs=/dev/dm-0
```

I used the added `find_mapper_device_name(module, …)` around the `os.path.realpath` on both places where you deleted then both path worked. After this I removed the `os.path.realpath` and both path still worked for me, but I assume that would not work for you.

Your udevadm command has in the last column the correct value for me too so it could be the best solution to add it.
```
root@rescue ~ # udevadm info --query=symlink --name=/dev/dm-0
disk/by-id/dm-name-decrypted disk/by-id/dm-uuid-CRYPT-LUKS1-184b8a210e2640c6b3375e220ef70f91-decrypted disk/by-id/lvm-pv-uuid-1LH5ae-k1Vk-sSYd-7vXd-2VRe-yjul-8tuYVE mapper/decrypted # <----
root@rescue ~ # udevadm info --query=symlink --name=/dev/mapper/decrypted
disk/by-id/dm-name-decrypted disk/by-id/dm-uuid-CRYPT-LUKS1-184b8a210e2640c6b3375e220ef70f91-decrypted disk/by-id/lvm-pv-uuid-1LH5ae-k1Vk-sSYd-7vXd-2VRe-yjul-8tuYVE mapper/decrypted # <----
```

Erouan50 committed:
Use udevadm command to get the real lvm path

lvm seems to get the device path from udev. So let's use udev
to to check if the device the is already in a pv, vg, etc...

Fix #43159

Erouan50 commented:
@NewJorg I added the command. I checked, and it seems to work. But if you can also confirm that it works in your side, it would be great. Better have a double check ;)

Thanks!

NewJorg commented:
@Erouan50 Using symlink and using realpath were both working for me with your fix.

Erouan50 commented:
Ok so let's keep the symlink/udevadm. Thank you!
